### DISCUSSION TITLE
Could the U54 paly as a master role?

### DISCUSSION THREAD

**Discussion by xiyanni** [polarfire-soc/polarfire-soc-documentation#102] [Bare metal embedded software] [answered]
If the user configure same as below, does it mean we don't need the E51 and nouse of the HSS?  
Do we need to program the eNVM?
Is U54_1 the master to start up and control U54_2,_3_4?
/////////////////////////////////////////////////////
#ifndef MPFS_HAL_FIRST_HART
#define MPFS_HAL_FIRST_HART  1
#endif

#ifndef MPFS_HAL_LAST_HART
#define MPFS_HAL_LAST_HART   4
#endif
/////////////////////////////////////////////////////

**Comment by hughbreslin** [comment, marked as answer]
Hi @xiyanni 

Using the configuration you shared the E51 would be held in wait for interrupt mode and unused. The U54_1 would configure the MSS and take the U54_2, U54_3 and U54_4 out of wait for interrupt once the system has started up.

You will need to use the HSS if your application is stored in an external non-volatile storage (e.g eMMC, SD card, QSPI) - if you're solely doing a bare metal application and its small enough to fit in the eNVM you could store your bare metal application in the eNVM instead of using the HSS and build the boot mode 1 configuration of the bare metal app. In any scenario you would need to program the eNVM. 

One point is that the HSS also provides other services such as scrubbing memories for ECC errors so it is useful to use.

**Reply by xiyanni** [reply]
What's the meaning"In any scenario you would need to program the eNVM."
When I don't use the E51 and do not use an external non-volatile storage, do I need to program eNVM? Why?

**Reply by hughbreslin** [reply]
My comment above was meant in a release scenario, if this is just for debug then you don't need to program the eNVM. If you want to run your application code on startup you will need to store it somewhere, which can be the eNVM if the application is small enough or else if its too large it could be stored somewhere off chip and loaded by the HSS.

**Reply by xiyanni** [reply]
Oh, yes. Anyway we need choose a storage for the application. So eNVM used is necessary in every condition. When we just do debug, where the application be stored?

**Reply by hughbreslin** [reply]
When you debug the application is downloaded to the destination memory, i.e if its built to target the LIM its downloaded to the LIM directly via the debugger. 

There are some potential issues that can come up with this, if the target memory is L2 scratchpad or DDR these need to be configured / trained first which means you will need to run an application to configure them - this can be done using the HSS by letting it boot up with your XML and then interrupting the boot and starting a debug session.

**Comment by xiyanni** [comment]
Another question: When user choose U54_1 as the master, why could not use the U54_1 to get the application from external non-volatile storage ?

**Reply by hughbreslin** [reply]
Theres nothing stopping a user setting up a U54 to load an application, the configuration we run with is to use the HSS on the E51 so it can stay active and running its house keeping services as its role is the monitor core and use the U54s as application cores.
